File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        //  ListNode *p=head,*q=nullptr,*r=nullptr;
        // while(p!=nullptr){
        //     r=q;
        //     q=p;
        //     p=p->next;
        //     q->next=r;
        // }
        // head=q;
        // return head;
        
        
        
        ListNode *prev=nullptr, *curr=nullptr, *front=head;
        while(front!=nullptr){
            prev=curr;
            curr=front;
            front=front->next;
            curr->next=prev;
        }
        head=curr;
        return head;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        //TC=O(N), SC=O(1).
        //Using three Pointers
        // ListNode prev=null, curr=null, front=head;
        // while(front!=null){
        //     prev=curr;
        //     curr=front;
        //     front=front.next;
        //     curr.next=prev;
        // }
        // head=curr;
        // return head;
        
        
        //TC=O(N), SC=O(N){recursion stack space.}
        //Using Recursion.
        if(head==null || head.next==null){
            return head;
        }
        ListNode newHead = reverseList(head.next);
        ListNode front = head.next;
        front.next=head;
        head.next=null;
        
        return newHead;
    }
}